File: solutions/c/roman-numerals/1/roman_numerals.c
#include "roman_numerals.h"
#include <string.h>
#include <malloc.h>

char *to_roman_numeral(unsigned int number){
    char numerals[] = {'M', 'D', 'C', 'L', 'X', 'V', 'I'};
    char buffer[15]={0};
    unsigned int steps[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
    int idx = 0;
    for (int i=0; i<13; i++) {
        while (number>=steps[i]) {
            number -= steps[i];
            if (0==i%2) {
                buffer[idx] = numerals[i/2];
            }
            else {
                buffer[idx++] = (steps[i]>100) ? 'C' : (steps[i]>10 ? 'X' : 'I');
                buffer[idx] = numerals[(i-1)/2];
            }
            idx++;
        }
    }
    unsigned int len = strlen(buffer);
    char *result = calloc(len,sizeof(char));
    strncpy(result, buffer, len);
    return result;
}

File: solutions/fortran/roman-numerals/1/roman_numerals.f90
module roman_numerals
  implicit none

character(len=2), dimension(13), parameter :: numerals = &
[character(len=2) :: 'M', 'CM', 'D', 'CD', 'C', 'XC', 'L', &
'XL', 'X', 'IX', 'V', 'IV', 'I']
integer, dimension(13), parameter :: steps = &
[1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]

contains

  function roman(num) result(s)
    integer, value :: num
    character(15) :: s
    integer :: i
    integer :: n

    s = ''
    n = num
    do i=1,13
        do while (n>=steps(i)) 
            n = n - steps(i)
            s = trim(s) // trim(numerals(i))
        end do
    end do

  end function roman

end module roman_numerals
